### PR TITLE
[MRG] Regression on pickling classes from the __main__ module

### PR DESCRIPTION
**Pull request by ogrisel**
This PR aims to fix the regression reported in #131 and that affects 0.4.2 and later. Right now there is is just a non-regression test to reproduce the problem.

I believe the regression was silently introduced by @pitrou's cleanups in #122 although I am not 100% sure yet because this code is quite complex.

### CODE REVIEW

File: tests/testutils.py [hunk 1]
```diff
@@ -1,7 +1,11 @@
 import sys
 import os
+import tempfile
 from subprocess import Popen
+from subprocess import check_output
 from subprocess import PIPE
+from subprocess import STDOUT
+from subprocess import CalledProcessError
```
Comment (pitrou) [inline tests/testutils.py]: Can you import all those names at once? :-)

File: tests/testutils.py [hunk 2]
```diff
@@ -64,9 +68,44 @@ def pickle_echo(stream_in=None, stream_out=None):
...
+    The script provided in the source code should return 0 and not print
+    anything on stderr or stdout.
+    """
+    fd, source_file = tempfile.mkstemp(suffix='_src_test_cloudpickle.py')
+    try:
+        with open(fd, 'wb') as f:
+            f.write(source_code.encode('utf-8'))
+
+        cmd = [sys.executable, source_file]
+        pythonpath = "{cwd}/tests:{cwd}".format(cwd=os.getcwd())
```
Comment (pitrou) [inline tests/testutils.py]: Can't you use `__file__` or something? `getcwd` sounds fragile.

**Comment by ogrisel** [comment]
@pitrou I have tried to make a minimal change to `save_global` to fix the tests but it does not feel right to me. Do you have a better suggestion?

**Comment by ogrisel** [comment]
The coverage decrease is caused by the new branch that can only be executed in subprocess calls and therefore not accounted in the report.

**Comment by pitrou** [comment]
> I have tried to make a minimal change to save_global to fix the tests but it does not feel right to me. 

I don't understand the fix. How does it work?

**Comment by ogrisel** [comment]
Apparently instances of `(type, types.ClassType)` from the `__main__` module need to go through `self.save_dynamic_class(obj)` instead of `Pickler.save_global(...)`.

**Comment by ogrisel** [comment]
I have pushed a simpler fix.

File: cloudpickle/cloudpickle.py [hunk 3]
```diff
@@ -628,12 +628,16 @@ def save_global(self, obj, name=None, pack=struct.pack):
         The name of this method is somewhat misleading: all types get
         dispatched here.
         """
+        if obj.__module__ == "__main__":
+            return self.save_dynamic_class(obj)
```
Comment (pitrou) [inline cloudpickle/cloudpickle.py]: Does this mean `save_dynamic_class` is also used for *functions*? If so, could you rename that method and update its docstring?

Comment (ssanderson) [reply cloudpickle/cloudpickle.py]: I don't think functions get dispatched here.  The dispatch entries for `save_global` are:

```python
     dispatch[type] = save_global
     dispatch[types.ClassType] = save_global
```

whereas function-likes are dispatched via:

```python
dispatch[types.FunctionType] = save_function
...
dispatch[types.MethodType] = save_instancemethod
...
dispatch[types.BuiltinFunctionType] = save_builtin_function
...
dispatch[classmethod] = save_classmethod
dispatch[staticmethod] = save_classmethod
```

I believe the name `save_global` is a holdover from the base `Pickler` class, which calls `self.save_global` in some base class methods that we call into via `super`, so renaming this isn't straightforward.

Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: I have pushed a new commit to simplify that function even further and all tests still pass. However, I am not sure we are not breaking edge cases in third party libraries and applications.

Comment (ogrisel) [reply cloudpickle/cloudpickle.py]: I have run the test suite of both loky and joblib against this branch all tests pass as well so I am pretty confident that the changes are fine.

Comment (pitrou) [reply cloudpickle/cloudpickle.py]: Can you check whether `save_dynamic_class` really gets a function as input and, if so, change the name or docstring?

Comment (ssanderson) [reply cloudpickle/cloudpickle.py]: @pitrou I'm pretty confident `save_dynamic_class` can never be called with a function as input.  It accesses `obj.__bases__` unconditionally, which doesn't exist on function objects.

**Comment by ogrisel** [comment]
No it does not. It's only for classes that cannot be looked up as attributes from global modules. In our tests 

- Classes that are defined in the local namespace of another function, method or class,
- Classes that are defined in the `__main__` module.

**Comment by ogrisel** [comment]
The name of the PR is misleading: the new non-regression tests call interactively defined functions from the `__main__` module but the actual problem was caused by the fact that interactively defined *classes* used by those functions could not be unpickled by attribute from module lookup.

**Comment by pitrou** [comment]
Ok, thank you for clarifying :-)

File: cloudpickle/cloudpickle.py [hunk 4]
```diff
@@ -628,18 +628,18 @@ def save_global(self, obj, name=None, pack=struct.pack):
...
-                    return self.save_reduce(_builtin_type, (_BUILTIN_TYPE_NAMES[obj],), obj=obj)
-
-            typ = type(obj)
-            if typ is not obj and isinstance(obj, (type, types.ClassType)):
-                return self.save_dynamic_class(obj)
+                    return self.save_reduce(
+                        _builtin_type, (_BUILTIN_TYPE_NAMES[obj],), obj=obj)
 
-            raise
+            return self.save_dynamic_class(obj)
```
Comment (pitrou) [inline cloudpickle/cloudpickle.py]: Does `save_dynamic_class` raise a readable exception or do we want to keep the `if` above?

**Comment by ogrisel** [comment]
Maybe it's good idea to keep the protection:

```python
>>> import cloudpickle
>>> import io
>>> p = cloudpickle.CloudPickler(io.BytesIO)
>>> p.save_dynamic_class(lambda x: x)
Fatal Python error: Cannot recover from stack overflow.

Thread 0x00007fc32d281700 (most recent call first):
  File "/home/ogrisel/.virtualenvs/py36/lib/python3.6/site-packages/IPython/core/history.py", line 764 in _writeout_input_cache
  File "/home/ogrisel/.virtualenvs/py36/lib/python3.6/site-packages/IPython/core/history.py", line 780 in writeout_cache
  File "/home/ogrisel/.virtualenvs/py36/lib/python3.6/site-packages/IPython/core/history.py", line 58 in needs_sqlite
  File "<decorator-gen-23>", line 2 in writeout_cache
  File "/home/ogrisel/.virtualenvs/py36/lib/python3.6/site-packages/IPython/core/history.py", line 834 in run
  File "/home/ogrisel/.virtualenvs/py36/lib/python3.6/site-packages/IPython/core/history.py", line 58 in needs_sqlite
  File "<decorator-gen-24>", line 2 in run
  File "/usr/lib/python3.6/threading.py", line 916 in _bootstrap_inner
  File "/usr/lib/python3.6/threading.py", line 884 in _bootstrap

Current thread 0x00007fc33466b700 (most recent call first):
  File "/usr/lib/python3.6/pickle.py", line 264 in _getattribute
  File "/usr/lib/python3.6/pickle.py", line 918 in save_global
  File "/home/ogrisel/code/cloudpickle/cloudpickle/cloudpickle.py", line 635 in save_global
  File "/home/ogrisel/code/cloudpickle/cloudpickle/cloudpickle.py", line 375 in save_function
  File "/usr/lib/python3.6/pickle.py", line 476 in save
  File "/home/ogrisel/code/cloudpickle/cloudpickle/cloudpickle.py", line 480 in save_dynamic_class
  File "/home/ogrisel/code/cloudpickle/cloudpickle/cloudpickle.py", line 642 in save_global
  File "/home/ogrisel/code/cloudpickle/cloudpickle/cloudpickle.py", line 375 in save_function
  File "/usr/lib/python3.6/pickle.py", line 476 in save
  File "/home/ogrisel/code/cloudpickle/cloudpickle/cloudpickle.py", line 480 in save_dynamic_class
  File "/home/ogrisel/code/cloudpickle/cloudpickle/cloudpickle.py", line 642 in save_global
  File "/home/ogrisel/code/cloudpickle/cloudpickle/cloudpickle.py", line 375 in save_function
  ...
Aborted (core dumped)
```

**Comment by ogrisel** [comment]
Actually the snippet from the previous comment is misleading, I passed the BytesIO class instead of an instance to build the pickler. With a correct pickler, the error message is fine I think:

```python
>>> import cloudpickle
>>> import io
>>> p = cloudpickle.CloudPickler(io.BytesIO())
>>> p.save_dynamic_class(lambda x: x)
Traceback (most recent call last):
  File "<ipython-input-8-4356e554f31b>", line 1, in <module>
    p.save_dynamic_class(lambda x: x)
  File "/home/ogrisel/code/cloudpickle/cloudpickle/cloudpickle.py", line 490, in save_dynamic_class
    # encountered while saving will point to the skeleton class.
AttributeError: 'function' object has no attribute '__bases__'
```

**Comment by pitrou** [comment]
But what was the previous message? It seems to me that Pickler generally raises PicklingError.

**Comment by ogrisel** [comment]
- `p.save_dynamic_classes(lambda x: x)` would raise the same error in 0.4.1 (as this part of the code has not changed).

- `p.save_global(lambda x: x)` would raise a `PicklingError: Can't pickle <function <lambda> at 0x7ff4d2ff97b8>` while not it raises the `AttributeError`.

However, I am not sure we should try to preserve the behavior of the private API on wrong inputs. What really matters is that it should raise a meaningful exception when calling the public API with wrong inputs and calling `loads(dumps(arbitrary_func))` should work on 0.5.2 as it used to work in 0.4.1 and not yield an exception.

**Comment by ogrisel** [comment]
@robertnishihara @iaroslav-ai could you please try this branch with your code to confirm that it fixes all your problems?

**Comment by ogrisel** [comment]
@pitrou it's true that the [upstream](https://github.com/python/cpython/blob/master/Lib/pickle.py#L920) `save_global` raises `PicklingError` when it cannot lookup such an object definition. I will restore this behavior in this branch.

However this case should never happen in cloudpickle as far as I know because cloudpickle has been written precisely to be able to pickle dynamic class and function definitions. I am not sure how to write a test for this.

**Comment by robertnishihara** [comment]
@ogrisel Thanks, I just tried it out and this seems to fix my problem.

**Comment by iaroslav-ai** [comment]
Yup that seems to solve it. Thanks @ogrisel !
